### ISSUE TITLE
Avoid Numpy for transforms

### ISSUE THREAD

**Issue by NevermindNilas** [LiheYoung/Depth-Anything#173]
Hi,

Currently my code goes as:
```py
        self.transform = Compose(
            [
                Resize(
                    width=518,
                    height=518,
                    resize_target=False,
                    keep_aspect_ratio=True,
                    ensure_multiple_of=14,
                    resize_method="lower_bound",
                    image_interpolation_method=cv2.INTER_CUBIC,
                ),
                NormalizeImage(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
                PrepareForNet(),
            ]
        )

    @torch.inference_mode()
    def processFrame(self, frame):
        try:
            # frame is a torch.uint8 tensor
            frame = (frame / 255.0).numpy()
            frame = self.transform({"image": frame})["image"]

            frame = torch.from_numpy(frame).unsqueeze(0).to(self.device)

            if self.half and self.isCudaAvailable:
                frame = frame.half()
            else:
                frame = frame.float()

            depth = self.model(frame)

            depth = F.interpolate(
                depth[None],
                (self.height, self.width),
                mode="bilinear",
                align_corners=False,
            )[0, 0]
            depth = (depth - depth.min()) / (depth.max() - depth.min()) * 255.0

            self.writeBuffer.write(depth)

        except Exception as e:
            logging.exception(f"Something went wrong while processing the frame, {e}")

        finally:
            self.semaphore.release()

```

Since Compose uses PIL / Numpy under the hood all of the operations are being offhanded to the CPU Instead of being done on the GPU. Would it be possible to keep the input/output as a torch tensor for as long as possible in order to avoid conversions to and from numpy and benefit as much as possible from the GPU's horse power?

Thank you.

**Comment by heyoeyo** [comment]
Since the `self.transform` part is just calling those 3 steps: `Resize`, `NormalizeImage` and `PrepareForNet` it's just a matter of re-implementing those steps using torch tensors. The code for each is defined the [depth_anything/util/transform.py](https://github.com/LiheYoung/Depth-Anything/blob/main/depth_anything/util/transform.py) script, so you can check that for reference of what processing needs to happen.

As the name suggests, the `Resize` transform is just [resizing](https://github.com/LiheYoung/Depth-Anything/blob/1e1c8d373ae6383ef6490a5c2eb5ef29fd085993/depth_anything/util/transform.py#L174-L178) the input image using `cv2.resize`. The equivalent operation using torch tensors would be [torch.nn.functional.interpolate](https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html) (or just `F.interpolate` which is already being used in the code you posted). You'd also have to copy the logic used to set the proper aspect ratio & multiples of 14 that the original resize operation was doing.

The `NormalizeImage` transform is just a simple [shifting/scaling operation](https://github.com/LiheYoung/Depth-Anything/blob/1e1c8d373ae6383ef6490a5c2eb5ef29fd085993/depth_anything/util/transform.py#L220), which can be done exactly the same way using torch tensors.

And then finally, the `PrepareForNet` transform looks like it's just [re-arranging the dimensions](https://github.com/LiheYoung/Depth-Anything/blob/1e1c8d373ae6383ef6490a5c2eb5ef29fd085993/depth_anything/util/transform.py#L233-L234) of the image from HxWxC to CxHxW and then making sure it's a float32 data type. The equivalent re-arranging operation with tensors is [permute](https://pytorch.org/docs/stable/generated/torch.permute.html).

**Comment by NevermindNilas** [comment]
Hi, you're totally right. I am sorry for being naive.

After converting all the nodes to Pytorch equivalents I've noticed a 50% bump in performance for CUDA and a 300% performance bump for TensorRT.

Thanks for the tips.

FWIW and future reference for others, most of the modifications can be seen [here](https://github.com/NevermindNilas/TheAnimeScripter/blob/main/src/depth/depth.py)
